File: Testing/www/js/testReferenceTypes.js
/* global Bridge */

"use strict";

Bridge.define('ClientTestLibrary.Class68', {
    x: 0,
    y: 1,
    test: function () {
        // Multiple local vars correctly
        var x = 1, y = 2;

        var z = x + y;
    }
});

Bridge.define('ClientTestLibrary.ClassA', {
    statics: {
        constructor: function () {
            ClientTestLibrary.ClassA.staticString = "Defined string";
            ClientTestLibrary.ClassA.staticInt = -340;
        },
        statitIntNotInitialized: 0,
        statitStringNotInitialized: null,
        staticInt: 0,
        staticString: null,
        CONST_CHAR: 81,
        CONST_DECIMAL: Bridge.Decimal("3.123456789324324324"),
        staticMethod1: function (i, s, d) {
            ClientTestLibrary.ClassA.statitIntNotInitialized = i;
            ClientTestLibrary.ClassA.statitStringNotInitialized = s;

            return Bridge.merge(new ClientTestLibrary.ClassA("constructor"), {
                setDoubleA: d
            } );
        },
        staticMethod2: function (p) {
            var i = Bridge.cast(p[0], Bridge.Int) + 1000;
            var s = Bridge.cast(p[1], String);
            var d = Bridge.cast(p[2], Number);

            return ClientTestLibrary.ClassA.staticMethod1(i, s, d);
        },
        tryParse: function (o, i) {
            i.v = 3;

            return true;
        },
        getDefaultInt: function () {
            return Bridge.getDefaultValue(Bridge.Int);
        }
    },
    data: null,
    config: {
        properties: {
            NumberA: 0,
            StringA: null,
            BoolA: false,
            DoubleA: 0,
            DecimalA: Bridge.Decimal(0.0)
        }
    },
    constructor: function () {
        this.setNumberA(10);
        this.setStringA("Str");
        this.setBoolA(true);
        this.setDoubleA(Number.POSITIVE_INFINITY);
        this.setDecimalA(Bridge.Decimal.MinusOne);
        this.setData(Bridge.merge(new ClientTestLibrary.ClassA.Aux1(), {
            setNumber: 700
        } ));
    },
    constructor$1: function (d) {
        this.$constructor();

        if (d === null)
            throw new Bridge.Exception("Related should not be null");

        this.setData(d);
    },
    constructor$2: function (p) {
        this.$constructor();

        if (p === null || p.length < 6) {
            throw new Bridge.Exception("Should pass six parameters");
        }

        if (Bridge.is(p[0], Bridge.Int)) {
            this.setNumberA(Bridge.cast(p[0], Bridge.Int));
        }

        if (Bridge.is(p[1], String)) {
            this.setStringA(Bridge.cast(p[1], String));
        }

        if (Bridge.is(p[2], Boolean)) {
            this.setBoolA(Bridge.cast(p[2], Boolean));
        }

        if (Bridge.is(p[3], Number)) {
            this.setDoubleA(Bridge.cast(p[3], Number));
        }

        if (Bridge.is(p[4], Bridge.Decimal)) {
            this.setDecimalA(Bridge.cast(p[4], Bridge.Decimal));
        }

        if (Bridge.is(p[5], ClientTestLibrary.ClassA.Aux1)) {
            this.setData(Bridge.cast(p[5], ClientTestLibrary.ClassA.Aux1));
        }
    },
    getData: function () {
        return this.data;
    },
    setData: function (value) {
        this.data = value;
    },
    method1: function () {
        var a1 = Bridge.merge(new ClientTestLibrary.ClassA.Aux1(), {
            setNumber: 1
        } );

        return Bridge.merge(new ClientTestLibrary.ClassA.Aux1(), {
            setNumber: 2,
            setRelated: a1
        } );
    },
    method2: function (a) {
        a.setRelated(a);
    },
    method3: function () {
        if (this.getData() !== null) {
            return this.getData().toString();
        }

        return "no data";
    },
    method4: function (i, add) {
        i = i + add;
        return i;
    },
    method5: function (i) {
        if (i === void 0) { i = -5; }
        return i;
    },
    method5$1: function (i, k) {
        if (i === void 0) { i = -50; }
        if (k === void 0) { k = -60; }
        return i + k;
    }
});

Bridge.define('ClientTestLibrary.ClassA.Aux1', {
    config: {
        properties: {
            Number: 0,
            Related: null
        }
    },
    toString: function () {
        return Bridge.String.format("{0} Has related {1}", this.getNumber(), this.getRelated() !== null ? this.getRelated().getNumber().toString() : "No");
    }
});

Bridge.define('ClientTestLibrary.TestReferenceTypes', {
    statics: {
        testInstanceConstructorsAndMethods: function (assert) {
            assert.expect(26);

            // Check parameterless constructor
            var a = new ClientTestLibrary.ClassA("constructor");

            // TEST
            assert.deepEqual(a.getNumberA(), 10, "NumberA 10");
            assert.deepEqual(a.getStringA(), "Str", "StringA Str");
            assert.deepEqual(a.getBoolA(), true, "BoolA true");
            assert.ok(a.getDoubleA() === Number.POSITIVE_INFINITY, "DoubleA Double.PositiveInfinity");
            assert.deepEqual(a.getDecimalA(), Bridge.Decimal(-1.0), "DecimalA Decimal.MinusOne");
            assert.ok(a.getData() !== null, "Data not null");
            assert.deepEqual(a.getData().getNumber(), 700, "Data.Number 700");

            // TEST
            // Check constructor with parameter
            assert.throws(ClientTestLibrary.TestSet1FailureHelper.testConstructor1Failure, "Related should not be null", "Related should not be null");

            // TEST
            // Check constructor with parameter
            assert.throws(ClientTestLibrary.TestSet1FailureHelper.testConstructor2Failure, "Should pass six parameters", "Should pass six parameters");

            a = new ClientTestLibrary.ClassA("constructor$2", [150, "151", true, 1.53, Bridge.Decimal(1.54), Bridge.merge(new ClientTestLibrary.ClassA.Aux1(), {
                setNumber: 155
            } )]);

            assert.deepEqual(a.getNumberA(), 150, "NumberA 150");
            assert.deepEqual(a.getStringA(), "151", "StringA 151");
            assert.deepEqual(a.getBoolA(), true, "BoolA true");
            assert.deepEqual(a.getDoubleA(), 1.53, "DoubleA Double.PositiveInfinity");
            assert.deepEqual(a.getDecimalA(), Bridge.Decimal(1.54), "DecimalA 154");
            assert.ok(a.getData() !== null, "Data not null");
            assert.deepEqual(a.getData().getNumber(), 155, "Data.Number 155");

            // TEST
            // Check instance methods
            var b = a.method1();

            assert.ok(b !== null, "b not null");
            assert.deepEqual(b.getNumber(), 2, "b Number 2");
            assert.ok(b.getRelated() !== null, "b.Related not null");
            assert.deepEqual(b.getRelated().getNumber(), 1, "b.Related Number 1");

            a.setData(b);
            assert.deepEqual(a.method3(), "2 Has related 1", "Method3 2 Has related 1");
            a.setData(null);
            assert.deepEqual(a.method3(), "no data", "Method3 no data");

            // TEST
            // Check [#68]
            var c68 = new ClientTestLibrary.Class68();

            assert.deepEqual(c68.x, 0, "c68.x 0");
            assert.deepEqual(c68.y, 1, "c68.y 1");

            // TEST
            // Check local vars do not get overridden by fields
            c68.test();

            assert.deepEqual(c68.x, 0, "c68.x 0");
            assert.deepEqual(c68.y, 1, "c68.y 1");
        },
        testStaticConstructorsAndMethods: function (assert) {
            assert.expect(13);

            // TEST
            // Check static fields initialization
            assert.deepEqual(ClientTestLibrary.ClassA.statitIntNotInitialized, 0, "#74 StatitInt not initialized");
            assert.deepEqual(ClientTestLibrary.ClassA.statitStringNotInitialized, null, "#74 StatitString not initialized");
            assert.deepEqual(ClientTestLibrary.ClassA.CONST_CHAR, 81, "#74 CONST_CHAR Q");
            assert.deepEqual(true, true, "#74 CONST_DECIMAL 3.123456789324324324m");

            // TEST
            // Check static constructor
            assert.deepEqual(ClientTestLibrary.ClassA.staticInt, -340, "StatitInt initialized");
            assert.deepEqual(ClientTestLibrary.ClassA.staticString, "Defined string", "StatitString initialized");

            // TEST
            // Check static methods
            var a = ClientTestLibrary.ClassA.staticMethod1(678, "ASD", Number.NaN);

            assert.deepEqual(ClientTestLibrary.ClassA.statitIntNotInitialized, 678, "StatitIntNotInitialized 678");
            assert.deepEqual(ClientTestLibrary.ClassA.statitStringNotInitialized, "ASD", "ClassA.StatitStringNotInitialized ASD");
            assert.deepEqual(a.getDoubleA(), Number.NaN, "DoubleA double.NaN");

            a = ClientTestLibrary.ClassA.staticMethod2([678, "QWE", 234]);
            assert.deepEqual(ClientTestLibrary.ClassA.statitIntNotInitialized, 1678, "StatitIntNotInitialized 1678");
            assert.deepEqual(ClientTestLibrary.ClassA.statitStringNotInitialized, "QWE", "ClassA.StatitStringNotInitialized QWE");
            assert.deepEqual(a.getDoubleA(), 234, "DoubleA 234");

            assert.throws(ClientTestLibrary.TestSet1FailureHelper.staticMethod2Failure, "Unable to cast type String to type Bridge.Int", "Cast exception should occur");
        },
        testMethodParameters: function (assert) {
            assert.expect(16);

            // Check default parameters
            var ra = new ClientTestLibrary.ClassA("constructor");
            var r = ra.method5(5);

            assert.deepEqual(r, 5, "r 5");
            r = ra.method5(15);
            assert.deepEqual(r, 15, "r 15");
            r = ra.method5$1(5, 6);
            assert.deepEqual(r, 11, "r 11");
            r = ra.method5$1(-50, 6);
            assert.deepEqual(r, -44, "r -44");

            // Check referencing did not change data
            var a = new ClientTestLibrary.ClassA("constructor");
            var b = a.method1();
            var c = b.getRelated();

            a.method2(b);
            assert.ok(b !== null, "b not null");
            assert.deepEqual(b.getNumber(), 2, "b Number 2");
            assert.ok(b.getRelated() !== null, "b.Related not null");
            assert.deepEqual(b.getRelated().getNumber(), 2, "b.Related Number 2");

            assert.ok(c !== null, "c not null");
            assert.deepEqual(c.getNumber(), 1, "c Number 1");
            assert.ok(c.getRelated() === null, "c.Related null");

            // Check value local parameter
            var input = 1;
            var result = a.method4(input, 4);

            assert.deepEqual(input, 1, "input 1");
            assert.deepEqual(result, 5, "result 5");

            // TEST
            // [#86]
            var di = ClientTestLibrary.ClassA.getDefaultInt();
            assert.deepEqual(di, 0, "di 0");

            // TEST
            // Check  "out parameter"
            // [#85]
            var i = { };
            var tryResult = ClientTestLibrary.ClassA.tryParse("", i);

            assert.ok(tryResult, "tryResult");
            assert.deepEqual(i.v, 3, "i 3");
        }
    }
});

Bridge.define('ClientTestLibrary.TestSet1FailureHelper', {
    statics: {
        testConstructor1Failure: function () {
            new ClientTestLibrary.ClassA("constructor$1", Bridge.cast(null, ClientTestLibrary.ClassA.Aux1));
        },
        testConstructor2Failure: function () {
            var t = new ClientTestLibrary.ClassA("constructor$2", [Bridge.Array.init(2, null)]);
        },
        staticMethod2Failure: function () {
            ClientTestLibrary.ClassA.staticMethod2(["1", "some string", "345.345435"]);
        }
    }
});

File: Testing/tests/client/BasicCSharp/TestReferenceTypes.cs
using Bridge;
using Bridge.Test;
using System;

namespace Bridge.ClientTest.BasicCSharp
{
    internal class ClassA
    {
        // TODO Add more types
        public static int StatitIntNotInitialized;

        public static string StatitStringNotInitialized;

        public static int StaticInt;
        public static string StaticString;
        public const char CONST_CHAR = 'Q';

        // TODO Add more to check big/small numbers, precision etc
        public const decimal CONST_DECIMAL = 3.123456789324324324m;

        // TODO Add more types
        public int NumberA
        {
            get;
            set;
        }

        public string StringA
        {
            get;
            set;
        }

        public bool BoolA
        {
            get;
            set;
        }

        public double DoubleA
        {
            get;
            set;
        }

        public decimal DecimalA
        {
            get;
            set;
        }

        private Aux1 data;

        public Aux1 Data
        {
            get
            {
                return data;
            }
            set
            {
                data = value;
            }
        }

        static ClassA()
        {
            ClassA.StaticString = "Defined string";
            ClassA.StaticInt = -340;
        }

        public ClassA()
        {
            this.NumberA = 10;
            this.StringA = "Str";
            this.BoolA = true;
            this.DoubleA = Double.PositiveInfinity;
            this.DecimalA = Decimal.MinusOne;
            this.Data = new Aux1()
            {
                Number = 700
            };
        }

        public ClassA(Aux1 d)
            : this()
        {
            if (d == null)
                throw new Exception("Related should not be null");

            this.Data = d;
        }

        // [#89]
        public ClassA(params object[] p)
            : this()
        {
            if (p == null || p.Length < 6)
            {
                throw new Exception("Should pass six parameters");
            }

            if (p[0] is int)
            {
                this.NumberA = (int)p[0];
            }

            if (p[1] is string)
            {
                this.StringA = (string)p[1];
            }

            if (p[2] is bool)
            {
                this.BoolA = (bool)p[2];
            }

            if (p[3] is double)
            {
                this.DoubleA = (double)p[3];
            }

            if (p[4] is decimal)
            {
                this.DecimalA = (decimal)p[4];
            }

            if (p[5] is Aux1)
            {
                this.Data = (Aux1)p[5];
            }
        }

        public ClassA.Aux1 Method1()
        {
            var a1 = new Aux1()
            {
                Number = 1
            };

            return new Aux1()
            {
                Number = 2,
                Related = a1
            };
        }

        public void Method2(ClassA.Aux1 a)
        {
            a.Related = a;
        }

        public string Method3()
        {
            if (this.Data != null)
            {
                return this.Data.ToString();
            }

            return "no data";
        }

        public int Method4(int i, int add)
        {
            i = i + add;
            return i;
        }

        public int Method5(int i = -5)
        {
            return i;
        }

        public int Method5(int i = -50, int k = -60)
        {
            return i + k;
        }

        public static ClassA StaticMethod1(int i, string s, double d)
        {
            ClassA.StatitIntNotInitialized = i;
            ClassA.StatitStringNotInitialized = s;

            return new ClassA()
            {
                DoubleA = d
            };
        }

        // [#89]
        public static ClassA StaticMethod2(params object[] p)
        {
            var i = (int)p[0] + 1000;
            var s = (string)p[1];
            var d = (double)p[2];

            return ClassA.StaticMethod1(i, s, d);
        }

        public static bool TryParse(object o, out int i)
        {
            i = 3;

            return true;
        }

        public static int GetDefaultInt()
        {
            return default(int);
        }
        // due to [#73] needs priority to be generated after the parent class
        // [Priority(-1)]
        public class Aux1
        {
            public int Number
            {
                get;
                set;
            }

            public Aux1 Related
            {
                get;
                set;
            }

            public override string ToString()
            {
                return string.Format("{0} Has related {1}", this.Number, this.Related != null ? this.Related.Number.ToString() : "No");
            }
        }
    }
    // [#68] Multiple field declaration renders incorrectly
    public class Class68
    {
        public int x, y = 1;

        public void Test()
        {
            // Multiple local vars correctly
            int x = 1, y = 2;

            var z = x + y;
        }
    }

    // Tests:
    // Reference type constructors, params method parameters, method overloading, nested classes, [FileName]
    // Full properties, short get/set properties, exceptions
    [Category(Constants.MODULE_BASIC_CSHARP)]
    [TestFixture(TestNameFormat = "Reference types - {0}")]
    public class TestReferenceTypes
    {
        // Check instance methods and constructors
        [Test(ExpectedCount = 26)]
        public static void TestInstanceConstructorsAndMethods()
        {
            // Check parameterless constructor
            var a = new ClassA();

            // TEST
            Assert.AreEqual(a.NumberA, 10, "NumberA 10");
            Assert.AreEqual(a.StringA, "Str", "StringA Str");
            Assert.AreEqual(a.BoolA, true, "BoolA true");
            Assert.True(a.DoubleA == Double.PositiveInfinity, "DoubleA Double.PositiveInfinity");
            Assert.AreDeepEqual(a.DecimalA, -1m, "DecimalA Decimal.MinusOne");
            Assert.True(a.Data != null, "Data not null");
            Assert.AreEqual(a.Data.Number, 700, "Data.Number 700");

            // TEST
            // Check constructor with parameter
            Assert.Throws(TestSet1FailureHelper.TestConstructor1Failure, "Related should not be null", "Related should not be null");

            // TEST
            // Check constructor with parameter
            Assert.Throws(TestSet1FailureHelper.TestConstructor2Failure, "Should pass six parameters", "Should pass six parameters");

            a = new ClassA(150, "151", true, 1.53d, 1.54m, new ClassA.Aux1()
            {
                Number = 155
            });

            Assert.AreEqual(a.NumberA, 150, "NumberA 150");
            Assert.AreEqual(a.StringA, "151", "StringA 151");
            Assert.AreEqual(a.BoolA, true, "BoolA true");
            Assert.AreEqual(a.DoubleA, 1.53, "DoubleA Double.PositiveInfinity");
            Assert.AreDeepEqual(a.DecimalA, 1.54m, "DecimalA 154");
            Assert.True(a.Data != null, "Data not null");
            Assert.AreEqual(a.Data.Number, 155, "Data.Number 155");

            // TEST
            // Check instance methods
            var b = a.Method1();

            Assert.True(b != null, "b not null");
            Assert.AreEqual(b.Number, 2, "b Number 2");
            Assert.True(b.Related != null, "b.Related not null");
            Assert.AreEqual(b.Related.Number, 1, "b.Related Number 1");

            a.Data = b;
            Assert.AreEqual(a.Method3(), "2 Has related 1", "Method3 2 Has related 1");
            a.Data = null;
            Assert.AreEqual(a.Method3(), "no data", "Method3 no data");

            // TEST
            // Check [#68]
            var c68 = new Class68();

            Assert.AreEqual(c68.x, 0, "c68.x 0");
            Assert.AreEqual(c68.y, 1, "c68.y 1");

            // TEST
            // Check local vars do not get overridden by fields
            c68.Test();

            Assert.AreEqual(c68.x, 0, "c68.x 0");
            Assert.AreEqual(c68.y, 1, "c68.y 1");
        }

        // Check static methods and constructor
        [Test(ExpectedCount = 13)]
        public static void TestStaticConstructorsAndMethods()
        {
            // TEST
            // Check static fields initialization
            Assert.AreEqual(ClassA.StatitIntNotInitialized, 0, "#74 StatitInt not initialized");
            Assert.AreEqual(ClassA.StatitStringNotInitialized, null, "#74 StatitString not initialized");
            Assert.AreEqual(ClassA.CONST_CHAR, 81, "#74 CONST_CHAR Q");
            Assert.AreEqual(ClassA.CONST_DECIMAL == 3.123456789324324324m, true, "#74 CONST_DECIMAL 3.123456789324324324m");

            // TEST
            // Check static constructor
            Assert.AreEqual(ClassA.StaticInt, -340, "StatitInt initialized");
            Assert.AreEqual(ClassA.StaticString, "Defined string", "StatitString initialized");

            // TEST
            // Check static methods
            var a = ClassA.StaticMethod1(678, "ASD", double.NaN);

            Assert.AreEqual(ClassA.StatitIntNotInitialized, 678, "StatitIntNotInitialized 678");
            Assert.AreEqual(ClassA.StatitStringNotInitialized, "ASD", "ClassA.StatitStringNotInitialized ASD");
            Assert.AreDeepEqual(a.DoubleA, double.NaN, "DoubleA double.NaN");

            a = ClassA.StaticMethod2((object)678, "QWE", 234);
            Assert.AreEqual(ClassA.StatitIntNotInitialized, 1678, "StatitIntNotInitialized 1678");
            Assert.AreEqual(ClassA.StatitStringNotInitialized, "QWE", "ClassA.StatitStringNotInitialized QWE");
            Assert.AreEqual(a.DoubleA, 234, "DoubleA 234");

            Assert.Throws(TestSet1FailureHelper.StaticMethod2Failure, "Unable to cast type String to type Bridge.Int", "Cast exception should occur");
        }

        // Check default parameters, method parameters, default values
        [Test(ExpectedCount = 16)]
        public static void TestMethodParameters()
        {
            // Check default parameters
            var ra = new ClassA();
            int r = ra.Method5(5);

            Assert.AreEqual(r, 5, "r 5");
            r = ra.Method5(i: 15);
            Assert.AreEqual(r, 15, "r 15");
            r = ra.Method5(5, 6);
            Assert.AreEqual(r, 11, "r 11");
            r = ra.Method5(k: 6);
            Assert.AreEqual(r, -44, "r -44");

            // Check referencing did not change data
            var a = new ClassA();
            var b = a.Method1();
            var c = b.Related;

            a.Method2(b);
            Assert.True(b != null, "b not null");
            Assert.AreEqual(b.Number, 2, "b Number 2");
            Assert.True(b.Related != null, "b.Related not null");
            Assert.AreEqual(b.Related.Number, 2, "b.Related Number 2");

            Assert.True(c != null, "c not null");
            Assert.AreEqual(c.Number, 1, "c Number 1");
            Assert.True(c.Related == null, "c.Related null");

            // Check value local parameter
            var input = 1;
            var result = a.Method4(input, 4);

            Assert.AreEqual(input, 1, "input 1");
            Assert.AreEqual(result, 5, "result 5");

            // TEST
            // [#86]
            var di = ClassA.GetDefaultInt();
            Assert.AreEqual(di, 0, "di 0");

            // TEST
            // Check  "out parameter"
            // [#85]
            int i;
            var tryResult = ClassA.TryParse("", out i);

            Assert.True(tryResult, "tryResult");
            Assert.AreEqual(i, 3, "i 3");
        }
    }

    public class TestSet1FailureHelper
    {
        // For testing exception throwing in constructors we need a separate method as constructors cannot be delegates
        public static void TestConstructor1Failure()
        {
            new ClassA((ClassA.Aux1)null);
        }

        public static void TestConstructor2Failure()
        {
            var t = new ClassA(new object[2]);
        }

        public static void StaticMethod2Failure()
        {
            ClassA.StaticMethod2("1", "some string", "345.345435");
        }
    }
}
